### PR TITLE
[@mantine/modals] Refactor modal manager & improve TypeScript types

### LINKED ISSUE #803: Context modals example gives error
### What package has an issue

@mantine/modals

### Describe the bug

I was trying to replicate the example at [Context modals](https://mantine.dev/others/modals/#context-modals) and I faced an error. 
The modal is displayed fine but gives an error.
```
Types of parameters '__0' and 'props' are incompatible.
    Type 'PropsWithChildren<ContextModalProps>' is not assignable to type 'ContextModalProps & { modalBody: string; }'.
```

### In which browser did the problem occur

Chrome

### If possible, please include a link to a codesandbox with the reproduced problem

https://codesandbox.io/s/mantine-context-modals-8nrfj?file=/pages/_app.tsx

### Do you know how to fix the issue

_No response_

### Are you willing to participate in fixing this issue and create a pull request with the fix

_No response_

### Possible fix

_No response_

### LINKED ISSUE #820: modal.openContextModal adds any extra props onto it's underlying `div`
Hi! So I noticed an issue with `modal.openContextModal` where any extra props you provide end up as attributes on the underlying `div`. React raises a warning about this saying `Warning: React does not recognize the `myPropName` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `mypropname` instead`

I've traced the issue back as follows:

- User calls `modal.openContextModal({ ..., categoryID })` with custom props, such as `categoryID`
- The [following snippet runs](https://github.com/mantinedev/mantine/blob/d414ad4f20ce4825d1f9e0e04ab59a7a7d1f40b2/src/mantine-modals/src/ModalsProvider.tsx#L86)
    - Note that `props` still contains the extra `categoryID` key
- The extra keys get sent to the Modal component as props by the [following snippet](https://github.com/mantinedev/mantine/blob/d414ad4f20ce4825d1f9e0e04ab59a7a7d1f40b2/src/mantine-modals/src/ModalsProvider.tsx#L138)
- The extra keys then get destructured as props onto a `<Box/>`, which will in turn destructure as props onto a `<div/>`
in the [following snippet](https://github.com/mantinedev/mantine/blob/d414ad4f20ce4825d1f9e0e04ab59a7a7d1f40b2/src/mantine-core/src/components/Modal/Modal.tsx#L169)

### PR DESCRIPTION
**Pull request by iDavidB**
Closes #803 
Closes #820

### CODE REVIEW

File: src/mantine-demos/src/demos/modals/context.tsx [hunk 1]
```diff
@@ -25,8 +27,10 @@ function Demo() {
   const openContextModal = () =>
     modals.openContextModal('demonstration', {
       title: 'Test modal from context',
-      modalBody:
-        'This modal was defined in ModalsProvider, you can open it anywhere in you app with useModals hook',
+      customProps: {
```
Comment (rtivital) [inline src/mantine-demos/src/demos/modals/context.tsx:30]: Let's rename customProps, it feels confusing, `innerProps` sounds a little better, I think.

Comment (iDavidB) [reply src/mantine-demos/src/demos/modals/context.tsx]: Done.

**Comment by iDavidB** [comment]
Found a bug in this setup. Working on a fix, not ready to merge.

**Comment by iDavidB** [comment]
This is ready to be reviewed again.

**Comment by rtivital** [comment]
Thanks!

**Comment by josiahayres** [comment]
Stoked to see this issue has been identified and fixed already 🥳

I am encountering [this](https://github.com/mantinedev/mantine/issues/803) issue with `4.2.12`, what version of Mantine will this fix be released in?

**Pull request merged**